### PR TITLE
Code coverage: Transaction service - overall %

### PR DESCRIPTION
**Pull request by Dimple16**
Screenshot 2023-02-13 at 1 33 02 PM

Note: Ignoring a few branch coverages

### CODE REVIEW

File: src/app/core/models/platform/platform-policy-expense.model.ts [hunk 1]
```diff
@@ -25,12 +25,16 @@ export interface PlatformPolicyExpense {
...
     longitude?: number;
   }[];
-  custom_fields?: [
-    {
-      name?: string;
-      value?: string;
-    }
-  ];
+  custom_fields?: {
+    id?: number;
```
Comment (rvab) [inline src/app/core/models/platform/platform-policy-expense.model.ts:29]: `id` isn't mandatory here?

File: src/app/core/models/platform/platform-policy-expense.model.ts [hunk 2]
```diff
@@ -25,12 +25,16 @@ export interface PlatformPolicyExpense {
...
+    mandatory?: boolean;
+    name: string;
+    options?: { label: string; value: string }[] | [];
+    placeholder?: string;
+    prefix?: string;
+    type?: string;
+    value: string | [];
```
Comment (rvab) [inline src/app/core/models/platform/platform-policy-expense.model.ts:36]: value can be anything no? Date, string, number. Either single item or an array of items

File: src/app/core/models/public-policy-expense.model.ts [hunk 3]
```diff
@@ -10,12 +10,16 @@ export interface PublicPolicyExpense {
...
-  ];
+  custom_properties: {
+    id?: number;
+    mandatory?: boolean;
+    name: string;
+    options?: { label: string; value: string }[] | [];
+    placeholder?: string;
+    prefix?: string;
+    type?: string;
+    value: string | [];
```
Comment (rvab) [inline src/app/core/models/public-policy-expense.model.ts:21]: same as above

File: src/app/core/models/txn-custom-properties.model.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+export interface TxnCustomProperties {
```
Comment (rvab) [inline src/app/core/models/txn-custom-properties.model.ts:1]: Can't we use this model in `platform-policy-expense.model.ts`?

File: src/app/core/services/transaction.service.spec.ts [hunk 5]
```diff
@@ -1452,4 +1464,75 @@ describe('TransactionService', () => {
...
+    const transactionId = 'txBRcjOg1spF';
+    const corporateCreditCardExpenseId = 'cccetzVpWd2Pgz';
+
+    transactionService.matchCCCExpense(transactionId, corporateCreditCardExpenseId).subscribe((res) => {
+      expect(res).toBeNull();
+      expect(apiService.post).toHaveBeenCalledWith('/transactions/match', {
+        transaction_id: transactionId,
+        corporate_credit_card_expense_id: corporateCreditCardExpenseId,
+      });
+      expect(apiService.post).toHaveBeenCalledTimes(1);
```
Comment (rvab) [inline src/app/core/services/transaction.service.spec.ts:1480]: `toHaveBeenCalledOnceWith`?

File: src/app/core/services/transaction.service.spec.ts [hunk 6]
```diff
@@ -1452,4 +1464,75 @@ describe('TransactionService', () => {
...
+    orgUserSettingsService.get.and.returnValue(of(orgUserSettingsData2));
+    spenderPlatformV1BetaApiService.post.and.returnValue(of(expensePolicyData));
+
+    transactionService.checkPolicy(platformPolicyExpenseData).subscribe((res) => {
+      expect(res).toEqual(expensePolicyData);
+      expect(spenderPlatformV1BetaApiService.post).toHaveBeenCalledOnceWith('/expenses/check_policies', {
+        data: platformPolicyExpenseData,
+      });
+      expect(orgUserSettingsService.get).toHaveBeenCalled();
+      expect(orgUserSettingsService.get).toHaveBeenCalledTimes(1);
```
Comment (rvab) [inline src/app/core/services/transaction.service.spec.ts:1495]: `toHaveBeenCalledTimes(1)` is sufficient. `toHaveBeenCalled` is not required.

File: src/app/core/services/transaction.service.spec.ts [hunk 7]
```diff
@@ -1452,4 +1464,75 @@ describe('TransactionService', () => {
...
+    // @ts-ignore
+    spyOn(transactionService, 'getTxnAccount').and.returnValue(of(txnAccountData));
+    timezoneService.convertAllDatesToProperLocale.and.returnValue(txnCustomPropertiesData2);
+    timezoneService.convertToUtc.and.returnValue(txnUtcDate);
+    utilityService.discardRedundantCharacters.and.returnValue(txnDataPayload);
+
+    transactionService.upsert(upsertTxnParam).subscribe((res) => {
+      expect(res).toEqual(txnData4);
+      expect(apiService.post).toHaveBeenCalledOnceWith('/transactions', txnDataPayload);
+      expect(orgUserSettingsService.get).toHaveBeenCalled();
```
Comment (rvab) [inline src/app/core/services/transaction.service.spec.ts:1528]: `toHaveBeenCalledTimes(1)`

**Review by rvab — CHANGES_REQUESTED** [review]
Few doubts

File: src/app/core/models/platform/platform-policy-expense.model.ts (hunk 1 above)
Comment (Dimple16) [reply src/app/core/models/platform/platform-policy-expense.model.ts:29]: no, few responses didn't have id

File: src/app/core/models/txn-custom-properties.model.ts (hunk 4 above)
Comment (Dimple16) [reply src/app/core/models/txn-custom-properties.model.ts:1]: platform-policy-expense model is corresponding to expense data, I'm adding an explicit model for custom properties, can reuse this in other places

File: src/app/core/models/platform/platform-policy-expense.model.ts (hunk 2 above)
Comment (Dimple16) [reply src/app/core/models/platform/platform-policy-expense.model.ts:36]: added value as `value: any | any[]`

**Review by rvab — APPROVED** [review]

**Pull request merged**